### ISSUE TITLE
dataset format

### ISSUE THREAD

**Issue by ZsZsZs25** [zhao-tong/Graph-Anomaly-Loss#1]
Thanks for your excellent work!
But what's the dataset format about these confused .pkl files? For example, why graph_u2u file in alpha dataset is not used  on GNN model but used in weibo dataset.
If there are some stupid mistakes I had made, please remind me! 
Thanks a lot!!!

**Comment by zhao-tong** [comment]
Hi, thanks for your interest! 
For the datafile formats, the `*graph_u2p.pkl `and `*graph_u2u.pkl` files are all in `scipy.sparse.csr_matrix` format. The label and feature files are all in `numpy.ndarray` format. They are all binary files and can be loaded with `pickle`. Sorry that the files now are kinda messy. I will organize them and add a readme file for data later.
The graph_u2u is actually just the inner product of the user-product adjacency matrix with the transpose of it self. It's used during the aggreagation (see Sec. 3.1 in our CIKM'20 paper for further details).
Please let me know if you have any other questions.

**Comment by ZsZsZs25** [comment]
> Hi, thanks for your interest!
> For the datafile formats, the `*graph_u2p.pkl `and `*graph_u2u.pkl` files are all in `scipy.sparse.csr_matrix` format. The label and feature files are all in `numpy.ndarray` format. They are all binary files and can be loaded with `pickle`. Sorry that the files now are kinda messy. I will organize them and add a readme file for data later.
> The graph_u2u is actually just the inner product of the user-product adjacency matrix with the transpose of it self. It's used during the aggreagation (see Sec. 3.1 in our CIKM'20 paper for further details).
> Please let me know if you have any other questions.

Thanks a lot for your early reply!
I think a readme file that describe the usage of these `.pkl` files will help a lot.
And I wonder why there are some difference when loading `alpha` and `weibo_s` dataset, such as `a_loss` is only used in `weibo_s` on GNN model.

**Comment by zhao-tong** [comment]
> And I wonder why there are some difference when loading alpha and weibo_s dataset, such as a_loss is only used in weibo_s on GNN model.

The data loading parts for alpha and weibo datasets should be roughly the same except the feature loading part. For the `a_loss` used in `weibo_s` on GNN model you mentioned, it was part of the testing codes I used when I was working on the paper. It should never be used and I'll remove those later when I clean-up the codes

**Comment by ZsZsZs25** [comment]
> > And I wonder why there are some difference when loading alpha and weibo_s dataset, such as a_loss is only used in weibo_s on GNN model.
> 
> The data loading parts for alpha and weibo datasets should be roughly the same except the feature loading part. For the `a_loss` used in `weibo_s` on GNN model you mentioned, it was part of the testing codes I used when I was working on the paper. It should never be used and I'll remove those later when I clean-up the codes

Thanks for your explanation!
So if I want to train my own dataset (including adjacency matrix, features, and labels), what else should I do? And in `get_cluster_aloss_pos_neg_nodes` function in `models.py` , `cluster_neighbors` should be loaded to compute $\Delta_{y_u}$, how can I generate that?

**Comment by zhao-tong** [comment]
When training with GAL, you would also need the results from an unsupervised method (e.g., `LockInfer`, `Fraudar`, `REV2`, etc). `[dataset]_labels_[method].pkl` saves the pickled prediction results of the unsupervised method, indicating which group/type/cluster each user belongs to. For dense block detection methods (e.g., `LockInfer`, `Fraudar`) that are able return multiple blocks/clusters, `[dataset]_clusters_[method].pkl` is a pickled list of list where each list contains the user ids in each block (overlapping is ok; other than the users in detected suspicious blocks, the remaining benign users is considered as the last block). Please refer to Section 3.4 (especially Eq. 24, 26, 27) for more details about this. 
The `cluster_neighbors` you mentioned are generated with `[dataset]_clusters_[method].pkl`. (https://github.com/zhao-tong/Graph-Anomaly-Loss/blob/master/src/dataCenter.py#L262)

**Comment by ZsZsZs25** [comment]
> When training with GAL, you would also need the results from an unsupervised method (e.g., `LockInfer`, `Fraudar`, `REV2`, etc). `[dataset]_labels_[method].pkl` saves the pickled prediction results of the unsupervised method, indicating which group/type/cluster each user belongs to. For dense block detection methods (e.g., `LockInfer`, `Fraudar`) that are able return multiple blocks/clusters, `[dataset]_clusters_[method].pkl` is a pickled list of list where each list contains the user ids in each block (overlapping is ok; other than the users in detected suspicious blocks, the remaining benign users is considered as the last block). Please refer to Section 3.4 (especially Eq. 24, 26, 27) for more details about this.
> The `cluster_neighbors` you mentioned are generated with `[dataset]_clusters_[method].pkl`. (https://github.com/zhao-tong/Graph-Anomaly-Loss/blob/master/src/dataCenter.py#L262)

Sorry to disturb you again due to my ignorance. 
I want to test GNN model with GAL, so firstly I use [fraudar](https://github.com/JunhaoWang/fraudar) code to catch dense blocks of `weibo` dataset, but unfortunately the results are not consistent with your resullts in [clusters_file](https://github.com/zhao-tong/Graph-Anomaly-Loss/blob/master/data/weibo_s/weibo_s_clusters_fraudar.pkl). What I did is just to store the egdes information about `graph_u2p` as the parameters of  `detectMultiple` function in [greedy.py](https://github.com/JunhaoWang/fraudar/blob/master/greedy.py).
Is  there any way eles to generate your `cluster` files?
Sorry to disturb again!!!

**Comment by zhao-tong** [comment]
I used the Fraudar package from Byan Hooi's website (modified a bit on `fraudarDetectMultiple`); I'm not sure if it is the same with the one you linked. Here's the script I used when generating those. When using on new datasets, `k` should be determined by the density of the blocks. 

```python
def fraudarDetectMultiple(M, detectFunc, numToDetect):
    # from the authors of Fraudar paper
    Mcur = copy.deepcopy(M).tolil()
    res = []
    scores = []
    for i in range(numToDetect):
        ((rowSet, colSet), score) = detectFunc(Mcur)
        res.append(rowSet)
        scores.append(score)
        (rs, cs) = Mcur.nonzero()
        for i in range(len(rs)):
            if rs[i] in rowSet and cs[i] in colSet:
                Mcur[rs[i], cs[i]] = 0 
    return res, np.array(scores)

res, scores = fraudarDetectMultiple(graph_u2p, aveDegree, 10)
k = 2
res_new = res[0:k+1]
scores_new = scores[0:k+1]
# normalize logits to (0,1]
logits = np.zeros(np.shape(graph_u2p)[0])
for i in range(len(res_new)):
    for x in res_new[i]:
        if logits[x] == 0:
            logits[x] = (scores_new[i]/np.max(scores_new))
# get splitting threshold
_, _, thresh = _eval(labels, logits)
# dump anomaly labels and clusters
logits[logits >= thresh] = 1
logits[logits < thresh] = 0
label_file = file_paths[ds]['fraudar']['a_label']
pickle.dump(logits, open(f'.{label_file}', 'wb'))
pred = set.union(*res)
users_left = set(range(np.shape(graph_u2p)[0]))
users_left = users_left - pred
res.append(users_left)
out_file = file_paths[ds]['fraudar']['a_cluster']
pickle.dump(res, open(f'.{out_file}', "wb"))

```

**Comment by ZsZsZs25** [comment]
Thanks a lot!!!
The code you posted is almost the same with mine. But still I can not generate the same results like yours, maybe caused by random seed.
But which metric is based to choose threshold?

**Comment by zhao-tong** [comment]
I used the threshold that gives me the best F1 score on the PR-curve. (btw, you should not be using the testing set when finding the threshold)

**Comment by zhao-tong** [comment]
Closing this issue as there's no follow-up questions.
